### DISCUSSION TITLE
Going from `a b c d` to `a b\nc d`

### DISCUSSION THREAD

**Discussion by pokey** [cursorless-dev/cursorless#1655] [Golf] [answered]
How do y'all go from `a b c d` to `a b\nc d` ? I do this constantly (rewrapping lines), so every syllable matters. Best I've got now is change leading cap newline .
EDIT: I guess I should rely on my editor to clean up trailing whitespace. At which point it's pre cap newline, which is probably optimal?
I'd love a custom command like flow cap, but that seems quite personal, and I don't quite see where to insert such a hook.

(this question was pasted from slack; originally posted by @josharian)

**Comment by pokey** [comment, marked as answer]
https://github.com/pokey/pokey_talon/blob/3a7ccd407be7b97104b99c7288f9771ce0c0db4e/apps/vscode/vscode.talon#L613-L616

so it would be `"break cap"`

For more information about how this works, see [the customization docs](https://www.cursorless.org/docs/user/customization/#cursorless-public-api)

**Reply by josharian** [reply]
This gift keeps on giving.

```
paste to <user.cursorless_target>:
    user.cursorless_command("setSelection", cursorless_target)
    user.vscode("execPaste")
paste before <user.cursorless_target>:
    user.cursorless_command("setSelectionBefore", cursorless_target)
    user.vscode("execPaste")
paste after <user.cursorless_target>:
    user.cursorless_command("setSelectionAfter", cursorless_target)
    user.vscode("execPaste")
```

**Reply by pokey** [reply]
Why not just use the built-in cursorless paste action?

**Reply by pokey** [reply]
Also, for the simple case of just running a custom VSCode action at a location, you're generally better served by using the custom actions csv, because that restores your cursor position afterward. See [the docs](https://www.cursorless.org/docs/user/customization/#experimental-cursorless-custom-ide-actions) and [this video](https://youtu.be/oWUJyDgz63k)

The full cursorless command api approach I use in my answer is only necessary if

- You need to use an action other than `setSelection`, and/or
- You need to do something else in addition to the custom action (eg hide the suggest widget), and/or
- You _don't_ want your cursor restored afterward.

**Reply by pokey** [reply]
Also note that the cursorless built-in paste action uses destinations rather than targets, so it's smart about inserting delimiters. Eg "paste before arg air" would insert a comma, and "paste after line air" would insert a newline. It's roughly equivalent to (and roughly implemented as) "pour target" followed by a paste, then restores cursor. It also sets the "that" mark so you can refer to the thing you just pasted

**Reply by josharian** [reply]
> Why not just use the built-in cursorless paste action?

Hahahah. Because my reading comprehension is apparently terrible. Or maybe my alphabetization? I looked for it on the cheatsheet and didn't find it and...didn't just try? Anyway, I will now!

**Comment by jmegner** [comment]
I do `"move tail <mark> after its line"`, which does not move the cursor like your break command does.  How would I do a talon command that does the same thing?  Is it even possible?  Something like...

```talon
split <user.cursorless_target>: user.cursorless_command("????????", cursorless_target)
```

It'd be really cool to do complicated cursorless commands like that.

**Reply by pokey** [reply]
Nope you'll need to wait for https://github.com/cursorless-dev/cursorless/issues/492 to do that

**Reply by jmegner** [reply]
Ah, thanks.
